src/Compilers/Core/Portable/CommandLine/AnalyzerConfigSet.cs

```diff
@@ -350,4 +370,205 @@ internal static bool TryParseSeverity(string value, out ReportDiagnostic severit
+            filteredConfigs.Insert(0, globalConfig);
+            return filteredConfigs.ToImmutableAndFree();
+
+            bool containsGlobalConfigs()
+            {
+                foreach (var config in analyzerConfigs)
+                {
+                    if (IsGlobalConfig(config))
+                    {
+                        return true;
+                    }
+                }
+                return false;
+            }
+        }
+
+        internal static bool IsGlobalConfig(AnalyzerConfig config) => config.GlobalSection.Properties.ContainsKey(GlobalKey);
+
+        private void MergeIntoGlobalConfig(AnalyzerConfig config)
+        {
+            MergeSection(config.PathToFile, config.GlobalSection);
+            foreach (var section in config.NamedSections)
+            {
+                MergeSection(config.PathToFile, section);
+            }
+        }
+
+        private AnalyzerConfig Build(out ImmutableArray<AnalyzerUnsetKey> unsetKeys)
+        {
+            unsetKeys = GetUnsetKeys();
+
+            Section globalSection = GetSection(string.Empty);
+            _values.Remove(string.Empty);
+
+            ArrayBuilder<Section> namedSectionBuilder = new ArrayBuilder<Section>(_values.Count);
+            foreach (var sectionName in _values.Keys)
+            {
+                namedSectionBuilder.Add(GetSection(sectionName));
+            }
+
+            AnalyzerConfig globalConfig = new AnalyzerConfig(globalSection, namedSectionBuilder.ToImmutableAndFree(), "/", isGlobal: true);
+            _values.Free();
+            return globalConfig;
+        }
+
+        private Section GetSection(string sectionName)
+        {
+            var dict = _values[sectionName];
+            var result = dict.ToImmutableDictionary(d => d.Key, d => d.Value.Item1, Section.PropertiesKeyComparer);
+            return new Section(sectionName, result);
+        }
+
+        private ImmutableArray<AnalyzerUnsetKey> GetUnsetKeys()
+        {
+            ArrayBuilder<AnalyzerUnsetKey> unsetKeys = ArrayBuilder<AnalyzerUnsetKey>.GetInstance();
+            foreach ((var section, var keys) in _duplicates)
+            {
+                bool isGlobalSection = string.IsNullOrWhiteSpace(section);
+                string sectionName = isGlobalSection ? "Global Section" : section;
+                foreach ((var keyName, var configs) in keys)
+                {
+                    unsetKeys.Add(new AnalyzerUnsetKey(keyName, sectionName, isGlobalSection, configs.ToImmutableAndFree()));
+                }
+                keys.Free();
+            }
+            _duplicates.Free();
+            return unsetKeys.ToImmutableAndFree();
+        }
+
+        private void MergeSection(string config, Section section)
+        {
+            if (!_values.ContainsKey(section.Name))
+            {
+                _values.Add(section.Name, ImmutableDictionary<string, (string, string)>.Empty.ToBuilder());
```

Comment:
Please use an explicit `StringComparer` here. Given this is comparing section names use, or define, on the `Section` type